### ISSUE TITLE
Please help to solve a problem with custom loss function

### ISSUE THREAD

**Issue by ntcnjdbxz** [keras-team/keras#8867]
Hello everyone,
I'm new in Keras and Neural networks. Nevertheless, I'm trying to implement a custom loss function for my own purpose. Assume I'm given a fat 2D array X=2xN. Where those 2 long vectors represent mixed audio signals.
My model now looks like that:
```
def generate_model():
    N=2
    model = Sequential()
    model.add(Conv1D(filters=N, kernel_size=(2), strides=1, activation="linear", use_bias=False,             
    kernel_initializer=unmixing_matrix_init, kernel_constraint=unit_norm(), input_shape=(None, 2))) 
    model.compile(loss=custom_loss, optimizer='sgd') 


X2 = np.expand_dims(X, axis=0)
Y2 = np.zeros((X2.shape))
model = generate_model() # Compile an neural network
model.fit(X2, Y2, epochs=100, batch_size=1)
model.evaluate(X2, Y2, batch_size=1, verbose=1)
X_output = model.predict(X2)
```

The problem now is in my loss function. So, I need to compute mutual info between those 2 audio signals in the loss function. That means I wanna use only y_pred in the loss function in such a way that CNN will estimate the best 4 weights in order to minimize/optimize the mutual info in my loss function. So, I want to implement BSS using CNN. Simple math:
X=A*S  (2xN=2x2 * 2xN) ---> S = W*X, where S - original channels, X - mixed channels, A - mixing matrix and W is 2x2 matrix with weights from CNN.
The problem now is how correctly define my model in order to be possible to separate my "2D" array into two "1D" arrays in the loss function. Then, to calculate mutual info using histogram or FFT of those 2 channels. Maybe it is possible to do so without splitting into 2 arrays using only one tensor?
Thanks in advance.

**Comment by gabrieldemarmiesse** [comment]
Hi!
Maybe I can help. Can you give the formula to compute the mutual information of two vectors? 
Plain math or numpy works for me. I'm just no familiar with mutual information. 
Then I can give you the code for the custom loss (if it's possible). 
Thanks!

**Comment by ntcnjdbxz** [comment]
Hi.
The pure math you can find in here [https://en.wikipedia.org/wiki/Mutual_information](url)

The code that I use in simple Python looks like this:

`
def entropy(x, bins):
  #Function to extimate the entropy of a signal (in a 1d array)
  #arguments: x: 1d array contains a signal
  #bins: # of bins for the histogram computation
  #returns:  entropy in terms of natural logarithm (nats)
  hist, binedges=np.histogram(x,bins)
  #print("hist=", hist)
  pdf=1.0*hist/np.sum(hist)
  entr= - np.sum(pdf*np.log(pdf+1e-6))
  return entr

def entropy2d(x,bins):
  #Function to extimate the joint entropy of 2 signals (in a 2d array)
  #arguments: x: 2d array, each of its 2 columns contains a signal
  #bins: # of bins for the histogram computation
  #returns: joint entropy in terms of natural logarithm (nats)

  hist2d, xedges, yedges=np.histogram2d(x[:,0],x[:,1],bins)
  pdf2d=1.0*hist2d/np.sum(hist2d)
  jointentropy= - np.sum(pdf2d*np.log(pdf2d+1e-6))
  return jointentropy
  
def mutualinformation(x,bins):
  #Computes the mutual information I(X;Y) of 2 signals in x
  #Arguments:
  #x: 2d array with 2 columns, 1 column for each signal
  #returns: mutual information absed on the natural logarithm (nats)
  #Computation based on https://en.wikipedia.org/wiki/Mutual_information
  #I(X;Y)= H(X)+H(Y)-H(X,Y)
  
  HX=entropy(x[:,0],bins) #entropy of X  
  HY=entropy(x[:,1],bins) #entropy of Y
  HXY=entropy2d(x,bins) #joint entropy of X and Y
  return HX+HY-HXY

def mutualinfocoeffs(coeffs,X):
  #computes the mutual information of 2 signals in array X columns, after multiplying with
  #an "crosstalk" matrix A, (multiplying each side with a factor and 
  #subtracting from the other)
  #to find the crosstalk matrix which produces the minimum mutual information
   
  A=np.array([[1.0, coeffs[0]], [coeffs[1], 1.0]])
  X_prime = np.dot(X,A)
  mutualinfo=mutualinformation(X_prime,100)
  return mutualinfo

# And I use minimization function to minimize this Mutual Info
X=np.dot(S,M)  # where S - vertical 2D array of pure audio channels, and M - mixing matrix 2x2
bounds=[(-2.0,2.0)]*2
coeffs_minimized  = opt.differential_evolution(mutualinfocoeffs, bounds, args=(X,), disp=True )
`

So, with Theano I tried to do following:
`
def custom_loss(y_true,y_pred):
    y_pred_1 = y_pred[:,:,0]
    y_pred_2 = y_pred[:,:,1]
`
Is it correct way to split y_pred into two arrays to use it in the future?
And second question - how to implement "pure" Python's "histogram2d" and "histogram" functions in Keras/Theano?

Thanks a lot.

**Comment by gabrieldemarmiesse** [comment]
Yes! This is the correct way of splitting tensors. It works on keras tensors.
 But for the function you're giving me, it involves computing the entropy, so computing an histogram. The problem is that computing an histogram is an operation that produces a gradient null nearly everwhere. So you can't optimise your network with stochastic gradient descend and backpropagation. Theorically, it's impossible. If you find a function that can approximate the entropy without producing a gradient of zero, it could be a good enough trick to save you.
To wrap up, it's theorically possible to do it, but your network would never update its weights. So you need to find a trick.
I hope it helps. Don't forget to close the issue if you're satisfied.
